### PR TITLE
8333750: GenShen: Only instantiate young/old generations in generational mode

### PR DESCRIPTION
**Pull request by earthling-amzn**
* Other modes should not instantiate or touch young/old generations.
* The generation sizer has been put in its own file and moved out of shHeap
* The logic for aging regions has been decoupled from non-generational final update refs

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8333750](https://bugs.openjdk.org/browse/JDK-8333750): GenShen: Only instantiate young/old generations in generational mode (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [d8215be6](https://git.openjdk.org/shenandoah/pull/444/files/d8215be627861900c0b3f0baf19507c0089220a6)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer) ⚠️ Review applies to [d8215be6](https://git.openjdk.org/shenandoah/pull/444/files/d8215be627861900c0b3f0baf19507c0089220a6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/444/head:pull/444` \
`$ git checkout pull/444`

Update a local copy of the PR: \
`$ git checkout pull/444` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/444/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 444`

View PR using the GUI difftool: \
`$ git pr show -t 444`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/444.diff">https://git.openjdk.org/shenandoah/pull/444.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/444#issuecomment-2153191886)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp [hunk 1]
```diff
@@ -111,6 +111,12 @@ bool ShenandoahOldGC::collect(GCCause::Cause cause) {
   // Complete marking under STW
   vmop_entry_final_mark();
 
+  if (_generation->is_concurrent_mark_in_progress()) {
+    assert(heap->cancelled_gc(), "Safepoint operation observed gc cancellation");
+    // GC may have been cancelled before final mark, but after the preceding cancellation check.
+    return false;
+  }
+
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:119]: This looks like it might be a bug fix, vs the rest of this PR, which is mostly isolation refactoring.  Should the bug fix be separated out into a different PR?

**Review by kdnilsen — APPROVED** [review]
Thanks for this refactoring.  All much deserved and important.  I wonder if the title of this PR should be a bit more general, since the refactoring appears to be more far reaching than just not instantiating young/old generations in non-generational mode.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp (hunk 1 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:119]: Yes, it is a bug fix. I'll pull it out into a separate PR.

Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:119]: Separate PR for this here: https://github.com/openjdk/shenandoah/pull/448

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp [hunk 2]
```diff
@@ -128,11 +128,16 @@ void ShenandoahHeapRegion::make_young_maybe() {
...
+       if (heap->mode()->is_generational()) {
+         if (is_old()) {
+           heap->old_generation()->decrement_affiliated_region_count();
+         }
        }
        set_affiliation(YOUNG_GENERATION);
-       ShenandoahHeap::heap()->young_generation()->increment_affiliated_region_count();
+       if (heap->mode()->is_generational()) {
+         heap->young_generation()->increment_affiliated_region_count();
+       }
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:140]: I assume the strict order here is on account of assertion checks in some of these methods? Otherwise, one might simplify it by moving the affiliation setting out from the sandwich of the count adjustments which are conditional on generational mode.

Ideally the count for young gen in non-generational mode would work for the generational case, and we could just adjust the old count like now and drop the adjustment of young count (lines 138-140) entirely, because they are covered by line 137.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp (hunk 1 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp:119]: Thanks for pulling this out separately. I do think this is subtle and we should (in the fullness of time) make this idiom a bit cleaner and less fragile and uniform in terms of the representation of the state machine that can be used by both the concurrent and the degenerate resumptions.

**Review by ysramakrishna — APPROVED** [review]
LGTM; left a few minor/optional comments, none of which require a re-review.

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp [hunk 3]
```diff
@@ -731,7 +730,7 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
       _heap->old_generation()->clear_cards_for(r);
       _heap->old_generation()->increment_affiliated_region_count();
     } else {
-      _heap->young_generation()->increment_affiliated_region_count();
+      _heap->generation_for(r->affiliation())->increment_affiliated_region_count();
     }
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:734]: Yes, that should be fine.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 4]
```diff
@@ -2054,9 +2036,12 @@ bool ShenandoahHeap::try_cancel_gc() {
...
 void ShenandoahHeap::cancel_concurrent_mark() {
-  _young_generation->cancel_marking();
-  _old_generation->cancel_marking();
-  _global_generation->cancel_marking();
+  if (mode()->is_generational()) {
+    young_generation()->cancel_marking();
+    old_generation()->cancel_marking();
+  }
+
+  global_generation()->cancel_marking();
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:2044]: I'll consider this for another PR. I think there are only a couple of places that would fit such a macro precisely and I'm not sure the obfuscating effects would be worth it.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 5]
```diff
@@ -2453,6 +2405,11 @@ void ShenandoahHeap::update_heap_region_states(bool concurrent) {
   }
 }
 
+void ShenandoahHeap::final_update_refs_do_regions() {
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:2408]: Yes, I thought the original name was too generic. `update_heap_region_states` doesn't indicate which phase of the collection it pertains to. I'll go with `final_update_refs_update_region_states` to match the timing phase.

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp (hunk 2 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:140]: We should be able to remove one of the mode checks. We don't maintain affiliated region counts in non-generational modes. There is a `TODO` for this: https://github.com/openjdk/shenandoah/blob/master/src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp#L872

I think this PR makes it easier to tackle that `TODO`, but it will be a separate PR.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp [hunk 6]
```diff
@@ -473,16 +488,22 @@ class ShenandoahHeap : public CollectedHeap {
...
+  ShenandoahYoungGeneration* young_generation()  const {
+    assert(mode()->is_generational(), "Young generation requires generational mode");
+    return _young_generation;
+  }
+
+  ShenandoahOldGeneration*   old_generation()    const {
+    assert(mode()->is_generational(), "Old generation requires generational mode");
+    return _old_generation;
+  }
+
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:505]: Yes, there are some usages internal to `ShenandoahHeap` that would be difficult to factor out without making certain "hot" methods `virtual` (on the allocation path, rebuilding free set, e.g.). I'm reluctant to do this for performance reasons, but perhaps it's worth the experiment?

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 7]
```diff
@@ -2378,58 +2363,25 @@ void ShenandoahHeap::update_heap_references(bool concurrent) {
...
-      // attempts to read the watermark will assert out (watermark should not be
-      // higher than top).
-      if (top > tams) {
-        // There have been allocations in this region since the start of the cycle.
-        // Any objects new to this region must not assimilate elevated age.
-        r->reset_age();
-      } else if (ShenandoahGenerationalHeap::heap()->is_aging_cycle()) {
-        r->increment_age();
+void ShenandoahSynchronizePinnedRegionStates::heap_region_do(ShenandoahHeapRegion* r) {
+  // Drop unnecessary "pinned" state from regions that does not have CP marks
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:2369]: I've simplified this comment.

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:734]: Yes, this should be fine.
